executor/insert_test.go

```diff
@@ -233,3 +235,34 @@ func (s *testSuite3) TestInsertZeroYear(c *C) {
 		`2000`,
 	))
 }
+
+func (s *testSuite3) TestAllowInvalidDates(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec(`use test`)
+	tk.MustExec(`drop table if exists t1;`)
+	tk.MustExec(`create table t1(d date);`)
+	tk.MustExec(`create table t2(d datetime);`)
+	tk.MustExec(`create table t3(d date);`)
+	tk.MustExec(`create table t4(d datetime);`)
+
+	runWithMode := func(mode string) {
+		inputs := []string{"0000-00-00", "2019-00-00", "2019-01-00", "2019-00-01", "2019-02-31"}
+		results := testkit.Rows(`0 0 0`, `2019 0 0`, `2019 1 0`, `2019 0 1`, `2019 2 31`)
+
```

Comment:
Maybe we should set the original sql_mode back after finishing this test, using defer in line 250.